### ISSUE TITLE
Spike in postgres connections when routes are added or updated

### ISSUE THREAD

**Issue by ganeshs** [Kong/kong#3634]
We are observing a spike in database connections whenever we add a route to a service or update an existing route. We are running kong version 0.14.0 on 8 (4 core ec2) machines with default configuration. 

On regular periods, the number of connections to database is 4 per node and we see 32 connections to the db overall. But during this issue, the connection goes beyond 2000 in the db. It lasts for less than a minute. During this period, the "active connections" spikes in kong and we see slowness in the calling systems. Below is our configuration, I have attached the kong connection metrics and rds connection metrics as well

<img width="908" alt="screen shot 2018-07-19 at 1 43 26 pm" src="https://user-images.githubusercontent.com/1185338/42930218-0b8456d6-8b5a-11e8-98bd-3e041476b205.png">
<img width="837" alt="screen shot 2018-07-19 at 1 44 23 pm" src="https://user-images.githubusercontent.com/1185338/42930219-0bd38f80-8b5a-11e8-9888-ebef56be9bd9.png">

`{"plugins":{"enabled_in_cluster":["prometheus","request-size-limiting","syslog"],"available_on_server":{"response-transformer":true,"oauth2":true,"acl":true,"correlation-id":true,"pre-function":true,"jwt":true,"cors":true,"ip-restriction":true,"basic-auth":true,"key-auth":true,"rate-limiting":true,"request-transformer":true,"http-log":true,"file-log":true,"hmac-auth":true,"ldap-auth":true,"datadog":true,"tcp-log":true,"zipkin":true,"post-function":true,"request-size-limiting":true,"bot-detection":true,"syslog":true,"loggly":true,"azure-functions":true,"udp-log":true,"response-ratelimiting":true,"aws-lambda":true,"statsd":true,"prometheus":true,"request-termination":true}},"tagline":"Welcome to kong","configuration":{"plugins":["bundled"],"admin_ssl_enabled":false,"lua_ssl_verify_depth":1,"trusted_ips":["0.0.0.0\/0"],"prefix":"\/usr\/local\/kong","loaded_plugins":{"response-transformer":true,"request-termination":true,"prometheus":true,"ip-restriction":true,"pre-function":true,"jwt":true,"cors":true,"statsd":true,"basic-auth":true,"key-auth":true,"ldap-auth":true,"aws-lambda":true,"http-log":true,"response-ratelimiting":true,"hmac-auth":true,"request-size-limiting":true,"datadog":true,"tcp-log":true,"zipkin":true,"post-function":true,"bot-detection":true,"acl":true,"loggly":true,"syslog":true,"azure-functions":true,"udp-log":true,"file-log":true,"request-transformer":true,"correlation-id":true,"rate-limiting":true,"oauth2":true},"cassandra_username":"kong","admin_ssl_cert_csr_default":"\/usr\/local\/kong\/ssl\/admin-kong-default.csr","dns_resolver":{},"pg_user":"kong","pg_password":"******","cassandra_data_centers":["dc1:2","dc2:3"],"nginx_admin_directives":{},"custom_plugins":["prometheus"],"pg_host":"kong-db.swiggy.prod","nginx_acc_logs":"\/usr\/local\/kong\/logs\/access.log","proxy_listen":["0.0.0.0:80"],"client_ssl_cert_default":"\/usr\/local\/kong\/ssl\/kong-default.crt","ssl_cert_csr_default":"\/usr\/local\/kong\/ssl\/kong-default.csr","dns_no_sync":false,"db_update_propagation":0,"nginx_err_logs":"\/usr\/local\/kong\/logs\/error.log","cassandra_port":9042,"dns_order":["A","CNAME","LAST","SRV"],"dns_error_ttl":1,"headers":["server_tokens","latency_tokens"],"dns_stale_ttl":30,"nginx_optimizations":true,"database":"postgres","pg_database":"db_kong","nginx_worker_processes":"auto","lua_package_cpath":"","admin_acc_logs":"\/usr\/local\/kong\/logs\/admin_access.log","lua_package_path":".\/?.lua;.\/?\/init.lua;","nginx_pid":"\/usr\/local\/kong\/pids\/nginx.pid","upstream_keepalive":60,"admin_access_log":"logs\/admin_access.log","client_ssl_cert_csr_default":"\/usr\/local\/kong\/ssl\/kong-default.csr","proxy_listeners":[{"ssl":false,"ip":"0.0.0.0","proxy_protocol":false,"port":80,"http2":false,"listener":"0.0.0.0:80"}],"proxy_ssl_enabled":false,"cassandra_contact_points":["127.0.0.1"],"enabled_headers":{"latency_tokens":true,"X-Kong-Proxy-Latency":true,"Via":true,"server_tokens":true,"Server":true,"X-Kong-Upstream-Latency":true,"X-Kong-Upstream-Status":false},"ssl_ciphers":"XXXXXXX","cassandra_ssl":false,"db_resurrect_ttl":30,"ssl_cert_key_default":"\/usr\/local\/kong\/ssl\/kong-default.key","cassandra_consistency":"ONE","db_cache_ttl":0,"admin_error_log":"logs\/error.log","pg_ssl_verify":false,"dns_not_found_ttl":30,"pg_ssl":false,"client_max_body_size":"0","client_ssl":false,"cassandra_repl_strategy":"SimpleStrategy","ssl_cipher_suite":"modern","lua_socket_pool_size":30,"db_update_frequency":5,"nginx_http_directives":[{"value":"prometheus_metrics 5m","name":"lua_shared_dict"}],"kong_env":"\/usr\/local\/kong\/.kong_env","cassandra_schema_consensus_timeout":10000,"dns_hostsfile":"\/etc\/hosts","admin_listeners":[{"ssl":false,"ip":"0.0.0.0","proxy_protocol":false,"port":8001,"http2":false,"listener":"0.0.0.0:8001"}],"cassandra_lb_policy":"RoundRobin","error_default_type":"text\/plain","real_ip_header":"X-Real-IP","admin_ssl_cert_key_default":"\/usr\/local\/kong\/ssl\/admin-kong-default.key","cassandra_ssl_verify":false,"proxy_error_log":"logs\/error.log","nginx_proxy_directives":{},"real_ip_recursive":"off","cassandra_repl_factor":1,"client_ssl_cert_key_default":"\/usr\/local\/kong\/ssl\/kong-default.key","nginx_daemon":"on","anonymous_reports":true,"cassandra_timeout":5000,"proxy_access_log":"logs\/access.log","pg_port":5432,"nginx_kong_conf":"\/usr\/local\/kong\/nginx-kong.conf","client_body_buffer_size":"400k","mem_cache_size":"128m","admin_ssl_cert_default":"\/usr\/local\/kong\/ssl\/admin-kong-default.crt","log_level":"notice","cassandra_keyspace":"kong","ssl_cert_default":"\/usr\/local\/kong\/ssl\/kong-default.crt","nginx_conf":"\/usr\/local\/kong\/nginx.conf","admin_listen":["0.0.0.0:8001"]},"version":"0.14.0","node_id":"f54f803b-6ebd-4bcf-aa9e-68f2157acb9d","lua_version":"LuaJIT 2.1.0-beta3","timers":{"pending":58,"running":1},"hostname":"kong-p-node-1"}`

**Comment by kikito** [comment]
On a multi-node setting like yours, when there is a change in one of the "core" entities (such as Services or Routes) it needs to be replicated to other nodes, and this is done via database queries. All nodes query the database periodically looking for changes.

One of the things that these event trigger is the re-creation of the routers. Routers are matching engines composed of all the routes in the database, and are cached per-worker. If you happen to have a lot of routes and/or services, this could be the cause of your spikes, since they all are reloaded from the database every time a new one is added, on every worker. Does this sound correct?

**Comment by ganeshs** [comment]
@kikito So, is it proportional to the number of concurrent calls being handled by nginx? Increasing our db connection limit for a not-so-frequent case of adding/updating routes seems overkill. Can the connection/cache be reused in this case via a background daemon?

**Comment by jeremyjpj0916** [comment]
This has come up in the forum too: https://discuss.konghq.com/t/route-caching-and-latency/840/4 , would be neat to solve as there is a bit of a risk if your running Kong really hot say 4000-5000 TPS and you add a route/service better hope your DB is ready for the heat 😆 . The idea of two cache primary and secondary (backup) and a switch on workers that uses the backup while while the new is being built out. Granted wait and run deployments at night all at once or some time outside critical hours and this can be avoided when your down to like 500-1000 TPS. Kong 15 improvement eh :P .

**Comment by ganeshs** [comment]
Uh. We will go with increasing the connections for time being. Hoping we will have a fix by 15. Thanks

**Comment by ganeshs** [comment]
@kikito @thibaultcha I have made a fix to not rebuild the router during the access phase. The approach I have tried is to trigger a worker_events.post to rebuild on all workers in the local node when there's a change in route. And also subscribe to cluster_events invalidations on 'router:version' to rebuild the router

https://github.com/Kong/kong/compare/master...ganeshs:master

Initial testing does seem to work. Would like to get an initial feedback on the approach before I build a full fledged fix. Thanks
